### PR TITLE
Feature/basic auth 2fa

### PR DESCRIPTION
**Pull request by zanaca**
Enabling basic auth at the frontend  and having an option to use 2FA tokens to secure the basic auth method

### CODE REVIEW

**Comment by chrisleekr** [comment]
Hey @zanaca 

This is awesome. 

Are you planning to update the frontend?

If you need any help, please let me know!

**Comment by zanaca** [comment]
I haven't thought about changing the frontend.  I just  saw as an "to do" on your list and  thought that a 2FA would be a nice way to control access over Localtunnel

**Comment by chrisleekr** [comment]
@zanaca Oh ok, I see. That should be good enough to protect access from the public for now.

Please see my review comments and write tests. Let me know when you are ready. :)
If you need any help, let me know.

File: app/server-frontend.js [hunk 1]
```diff
@@ -1,15 +1,58 @@
...
+  if (config.frontend_auth.mode_otp && OTP_SECRET) {
+    const passwordPlain = basicAuth.safeCompare(password.slice(0, -6), config.frontend_auth.password);
+    const password2fa = Boolean(otp.verifyToken(OTP_SECRET, password.replace(config.frontend_auth.password, '')))
+
+    passwordOk = passwordPlain && password2fa;
+  } else {
+    passwordOk = basicAuth.safeCompare(password, config.frontend_auth.password);
+  }
+
+  return usernameOk & passwordOk;
```
Comment (chrisleekr) [inline app/server-frontend.js:36]: Should be `&&` not `&`

File: app/server-frontend.js [hunk 2]
```diff
@@ -1,15 +1,58 @@
...
   logger.info({ config }, `API ${config.get('mode')} frontend started on`);
 
   const app = express();
+  if (config.frontend_auth) {
+    const baOptions = {
+      authorizer: basicAuthChallenge,
+      challenge: true,
+      realm: 'Binance Trading Bot'
+    };
+    if (config.frontend_auth.mode_otp) {
```
Comment (chrisleekr) [inline app/server-frontend.js:50]: I have tried with OTP mode on, but without OTP entered, I still can access the frontend. 

image

image

image

Comment (chrisleekr) [inline app/server-frontend.js:50]: I did generate `.2fa_secret` and set the configuration as below:
image

However, I still can access without adding OTP code to the password

image

image

Comment (zanaca) [reply app/server-frontend.js:50]: Nothing.. it's a bug about the password/2fa length

Comment (zanaca) [reply app/server-frontend.js]: working now

**Comment by chrisleekr** [comment]
Hey @zanaca 

I did some code change. Could you check and see you are happy with the changes?
And I am writing test code as well.

**Comment by chrisleekr** [comment]
@zanaca 
I have added test codes as well. 
Can you review and let me know if you are happy with the changes?
Then I will merge in.

File: app/frontend/basic-auth.js [hunk 3]
```diff
@@ -69,7 +57,7 @@ const setBasicAuth = (app, funcLogger) => {
       logger.info('2FA authentication is enabled. Add 2FA option.');
       basicAuthOptions.realm += ' - 2FA';
     }
-
+    console.log('basicAuthOptions => ', basicAuthOptions);
```
Comment (zanaca) [inline app/frontend/basic-auth.js]: Floating console.log

File: app/frontend/basic-auth.js [hunk 4]
```diff
@@ -21,25 +15,22 @@ const basicAuthChallenge = (usernamePlain, passwordPlain) => {
...
+      authPassword
     );
 
     const validPassword2FA = Boolean(
-      otp.verifyToken(SECRET_2FA, passwordPlain.replace(authPassword, ''))
+      otp.verifyToken(SECRET_2FA, plainPassword.replace(authPassword, ''))
     );
 
-    validPassword = validPasswordPlain && validPassword2FA;
+    validPassword = validPlainPassword && validPassword2FA;
```
Comment (zanaca) [inline app/frontend/basic-auth.js:30]: 👍

**Review by zanaca — COMMENTED** [review]
Better use of config.get!  I am not used to it but it is way better for reading! 

Just noticed a console.log. 

PS: I cannot aprove the changes 👍

File: app/frontend/basic-auth.js (hunk 3 above)
Comment (chrisleekr) [reply app/frontend/basic-auth.js]: Oh I think you reviewed outdated code. Please check again. :)

File: app/frontend/basic-auth.js (hunk 4 above)
Comment (chrisleekr) [reply app/frontend/basic-auth.js:30]: Same thing, please check the updated code.

**Comment by chrisleekr** [comment]
@zanaca 
> PS: I cannot aprove the changes

Ah, just simply review the code and let me know you are happy or not :)
Then I will merge in.

**Comment by ganesshkumar** [comment]
Can we shit this, please? xD

File: script/generate-2fa-secret.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env node
+/* eslint-disable no-console */
+const otp = require('node-2fa');
+const config = require('config');
+const fs = require('fs');
+const qrcode = require('qrcode-terminal');
+
+const username = config.get('frontend.auth.username');
+const password = config.get('fonrtend.auth.password');
```
Comment (DreamN) [inline script/generate-2fa-secret.js:9]: Hi, I've just found that there's an incorrect typo on this line. :)

**Comment by Conight** [comment]
The best way to use `basicauth` is using reverse proxy like `caddy` or `nginx` and add some configure by yourself.
No need to change code.

My caddy server config example:
```Caddyfile
trading.example.com {
    reverse_proxy 127.0.0.1:8080
    basicauth * {
        Username EncodedPassword
    }
}
```

- Caddy doc click [here](https://caddyserver.com/docs/caddyfile/directives/basicauth).
- Nginx doc click [here](https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-http-basic-authentication/).

**Comment by chrisleekr** [comment]
I actually want to add this feature in a different way.

1. Option to secure the whole frontend by not showing any data until authenticate. But not via basic auth.
2. Option to display data in the frontend but secure configuration change. Need to authenticate to change the configuration.

I just don't have much time to implement but yeah, one day it will come.

**Comment by Conight** [comment]
> I actually want to add this feature in a different way.
> 
> 1. Option to secure the whole frontend by not showing any data until authenticate. But not via basic auth.
> 2. Option to display data in the frontend but secure configuration change. Need to authenticate to change the configuration.
> 
> I just don't have much time to implement but yeah, one day it will come.

Yeah, I just give people a way who want use trading with auth immediately.

Thx for your excellent work!

**Comment by braverynew** [comment]
> The best way to use `basicauth` is using reverse proxy like `caddy` or `nginx` and add some configure by yourself.
> No need to change code.
> 
> My caddy server config example:
> 
> ```
> trading.example.com {
>     reverse_proxy 127.0.0.1:8080
>     basicauth * {
>         Username EncodedPassword
>     }
> }
> ```
> 
> * Caddy doc click [here](https://caddyserver.com/docs/caddyfile/directives/basicauth).
> * Nginx doc click [here](https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-http-basic-authentication/).

It's very simple way!!!. Thank a lot

**Comment by chrisleekr** [comment]
This PR is closed in favour of #260.

**Pull request closed without merging**